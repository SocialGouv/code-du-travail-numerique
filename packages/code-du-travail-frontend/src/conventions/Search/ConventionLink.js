import { formatIdcc } from "@cdt/data";
import slugify from "@socialgouv/cdtn-slugify";
import { Button, Paragraph, theme } from "@socialgouv/cdtn-ui";
import Link from "next/link";
import { useRouter } from "next/router";
import React from "react";
import styled from "styled-components";

import { matopush } from "../../piwik";

// following @cdt/data/indexing/cdtnDocuments.js slug rules
const getConventionSlug = (convention) =>
  slugify(`${convention.num}-${convention.shortTitle}`.substring(0, 80));

export const ConventionLink = ({
  convention,
  isFirst,
  onClick,
  small = false,
}) => {
  const { num, shortTitle, highlight } = convention;
  const router = useRouter();

  const clickHandler = () => {
    matopush(["trackEvent", "cc_select", router.asPath, shortTitle]);
    onClick && onClick(convention);
  };

  const commonProps = {
    isFirst,
    onClick: clickHandler,
    small,
  };

  return onClick ? (
    <StyledLink as={Button} variant="navLink" {...commonProps}>
      {shortTitle} <IDCC>(IDCC {formatIdcc(num)})</IDCC>
      {highlight && (
        <Paragraph variant="altText">{highlight.searchInfo}</Paragraph>
      )}
    </StyledLink>
  ) : (
    <Link
      href={`/convention-collective/${getConventionSlug({
        num,
        shortTitle,
      })}`}
      passHref
    >
      <StyledLink {...commonProps}>
        {shortTitle} <IDCC>(IDCC {formatIdcc(num)})</IDCC>
        {highlight && (
          <Paragraph variant="altText">{highlight.searchInfo}</Paragraph>
        )}
      </StyledLink>
    </Link>
  );
};

const { spacings } = theme;

const StyledLink = styled.a`
  display: block;
  padding: ${({ small }) => (small ? spacings.xsmall : spacings.base)} 0;
  ${({ isFirst }) => isFirst && "padding-top: 0;"}
  color: ${({ theme }) => theme.secondary};
  font-weight: 600;
  text-align: left;
  text-decoration: none;
`;

const IDCC = styled.span`
  font-weight: normal;
`;
