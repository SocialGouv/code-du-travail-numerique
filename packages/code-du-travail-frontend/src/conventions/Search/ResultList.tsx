import { Heading, theme, ViewMore, Wrapper } from "@socialgouv/cdtn-ui";
import React from "react";
import { useUIDSeed } from "react-uid";
import styled from "styled-components";

type Props = {
  query: string;
  title: string;
  items: JSX.Element[];
};

export const ResultList = ({ items, query, title }: Props): JSX.Element => {
  const seedId = useUIDSeed();
  return (
    <StyledWrapper variant="light">
      <Heading>
        {`${title} (${items.length}`}&nbsp;
        {`rÃ©sultat${items.length > 1 ? "s" : ""})`}
      </Heading>
      <ViewMore query={query} stepSize={7}>
        {items.map((item) => (
          <StyledListItem key={seedId(item)}>{item}</StyledListItem>
        ))}
      </ViewMore>
    </StyledWrapper>
  );
};

const { box, spacings } = theme;

const StyledWrapper = styled(Wrapper)`
  & + & {
    margin-top: ${spacings.base};
  }
`;

const StyledListItem = styled.li`
  & + & {
    border-top: ${({ theme }) => box.border(theme.border)};
  }

  &:last-of-type {
    & > *:last-child {
      margin-bottom: 0;
      padding-bottom: 0;
    }
  }
`;
