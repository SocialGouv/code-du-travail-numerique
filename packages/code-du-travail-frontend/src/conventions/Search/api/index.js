import { parseIdcc } from "@socialgouv/modeles-social";

import { searchAgreement } from "./agreement.service";
import {
  searchEntrepriseByName,
  searchEntrepriseBySiren,
  searchEntrepriseBySiret,
} from "./entreprise.service";
import getQueryType from "./getQueryType";

// build a result list based on query type
export const getResults = async (query) => {
  const trimmedQuery = query.trim();
  const type = getQueryType(query);

  let conventions = [];
  let entreprises = [];

  if (type === "text") {
    [conventions, entreprises] = await Promise.all([
      searchAgreement(trimmedQuery),
      searchEntrepriseByName(trimmedQuery).then((entreprises) =>
        entreprises.filter(
          (entreprise) =>
            entreprise.conventions && entreprise.conventions.length
        )
      ),
    ]);
  } else if (type === "siren") {
    entreprises = await searchEntrepriseBySiren(query.replace(/[\s .-]/g, ""));
  } else if (type === "siret") {
    entreprises = await searchEntrepriseBySiret(query.replace(/[\s .-]/g, ""));
  } else if (type === "idcc") {
    const matches = await searchAgreement(parseIdcc(trimmedQuery));

    if (matches && matches.length) {
      const perfectMatch = matches.find(
        (match) => parseIdcc(match.num) === parseIdcc(trimmedQuery)
      );
      conventions = perfectMatch ? [perfectMatch] : matches.slice(0, 5);
    }
  } else {
    return null;
  }
  return {
    conventions,
    entreprises: entreprises.filter(
      // we might want to remove this in a near future
      (entreprise) => !entreprise.closed && entreprise.conventions.length
    ),
  };
};
