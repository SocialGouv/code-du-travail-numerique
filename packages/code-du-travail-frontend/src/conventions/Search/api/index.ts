import { parseIdcc } from "@socialgouv/modeles-social";

import { searchAgreement } from "./agreement.service";
import getQueryType, { QueryType } from "./getQueryType";
import { Enterprise, searchEnterprises } from "./enterprises.service";
import { Agreement } from "../../../outils/types";

export type ResultType = {
  conventions: Agreement[];
  entreprises: Enterprise[];
} | null;
// build a result list based on query type
export const getResults = async (query: string): Promise<ResultType | null> => {
  const trimmedQuery = query.trim();
  const type = getQueryType(query);

  let conventions: Agreement[] = [];
  let entreprises: Enterprise[] = [];

  switch (type) {
    case QueryType.TEXT:
      [conventions, entreprises] = await Promise.all([
        searchAgreement(trimmedQuery),
        searchEnterprises({ query: trimmedQuery }).then((entreprises) =>
          entreprises.filter(
            (entreprise) =>
              entreprise.conventions && entreprise.conventions.length,
          ),
        ),
      ]);
      break;
    case QueryType.IDCC:
      const matches = await searchAgreement(trimmedQuery);

      if (matches && matches.length) {
        const perfectMatch = matches.find(
          (match) => match.num === parseIdcc(trimmedQuery),
        );
        conventions = perfectMatch ? [perfectMatch] : matches.slice(0, 5);
      }
      break;
    default:
      return null;
  }
  return {
    conventions,
    entreprises,
  };
};
