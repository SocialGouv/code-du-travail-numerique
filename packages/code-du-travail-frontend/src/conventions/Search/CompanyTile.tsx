import { theme } from "@socialgouv/cdtn-ui";
import React from "react";
import styled from "styled-components";

import { ConventionLink } from "./ConventionLink";
import { Agreement } from "../../outils/types";

type Props = {
  conventions: Agreement[];
  label: string;
  onClick: (agreement: Agreement) => void;
};

export const CompanyTile = ({
  conventions = [],
  label,
  onClick,
}: Props): JSX.Element => (
  <>
    <Entreprise>{label}</Entreprise>
    {conventions.length ? (
      conventions.map((convention) => (
        <ConventionLink
          convention={convention}
          small
          key={convention.num}
          onClick={onClick}
        />
      ))
    ) : (
      <Div>Aucune convention collective connue pour cette entreprise</Div>
    )}
  </>
);

const { spacings } = theme;

const Entreprise = styled.div`
  margin: ${spacings.xsmall} 0 0 0;
  font-weight: bold;
`;

const Div = styled.div`
  margin-bottom: ${spacings.xsmall};
`;
