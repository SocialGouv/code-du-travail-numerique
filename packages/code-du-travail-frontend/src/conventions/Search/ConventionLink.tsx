import { formatIdcc } from "@socialgouv/modeles-social";
import { Button, Paragraph, theme } from "@socialgouv/cdtn-ui";
import { push as matopush } from "@socialgouv/matomo-next";
import Link from "next/link";
import { useRouter } from "next/router";
import React from "react";
import styled from "styled-components";
import { Agreement } from "./api/type";

type Props = {
  convention: Agreement;
  isFirst?: boolean;
  onClick?: (agreement: Agreement) => void;
  small?: boolean;
};

export const ConventionLink = ({
  convention,
  isFirst,
  onClick,
  small = false,
}: Props): JSX.Element => {
  const { num, shortTitle, highlight } = convention;
  const router = useRouter();

  const clickHandler = () => {
    matopush(["trackEvent", "cc_select", router.asPath, shortTitle]);
    onClick && onClick(convention);
  };

  const commonProps = {
    isFirst,
    onClick: clickHandler,
    small,
  };

  return onClick ? (
    <StyledLink as={Button} variant="navLink" {...commonProps}>
      {shortTitle} <IDCC>(IDCC {formatIdcc(num)})</IDCC>
      {highlight && highlight.searchInfo && (
        <Paragraph variant="altText" noMargin>
          {highlight.searchInfo}
        </Paragraph>
      )}
    </StyledLink>
  ) : (
    <Link
      href={`/convention-collective/${convention.slug}`}
      passHref
      legacyBehavior
    >
      <StyledLink {...commonProps}>
        {shortTitle} <IDCC>(IDCC {formatIdcc(num)})</IDCC>
        {highlight && highlight.searchInfo && (
          <Paragraph variant="altText" noMargin>
            {highlight.searchInfo}
          </Paragraph>
        )}
      </StyledLink>
    </Link>
  );
};

const { spacings } = theme;

const StyledLink = styled.a`
  display: block;
  padding: ${({ small }) => (small ? spacings.xsmall : spacings.base)} 0;
  ${({ isFirst }) => isFirst && "padding-top: 0;"}
  color: ${({ theme }) => theme.secondary};
  font-weight: 600;
  text-align: left;
  text-decoration: none;
`;

const IDCC = styled.span`
  font-weight: normal;
`;
