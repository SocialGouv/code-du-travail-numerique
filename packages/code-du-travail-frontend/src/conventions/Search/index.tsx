import { Button, Input, Label, Paragraph, theme } from "@socialgouv/cdtn-ui";
import { push as matopush } from "@socialgouv/matomo-next";
import debounce from "debounce-promise";
import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
import styled from "styled-components";
import { v4 as generateUUID } from "uuid";

import { CompanyTile } from "./CompanyTile";
import { ConventionLink } from "./ConventionLink";
import { HelpModal } from "./HelpModal";
import { ResultList } from "./ResultList";
import useSearchCC, { Status } from "./searchHook";
import Spinner from "react-svg-spinner";
import { Agreement } from "../../outils/types";

const trackInput = debounce((query, path, trackingUID) => {
  if (query.length > 1) {
    matopush(["trackEvent", "cc_search", path, `${trackingUID} : ${query}`]);
  }
}, 2000);

type Props = {
  onSelectConvention?: (agreement: Agreement) => void;
};

const Search = ({ onSelectConvention }: Props): JSX.Element => {
  const router = useRouter();
  const [query, setQuery] = useState<string>("");
  const [trackingUID, setTrackingUID] = useState<string>("");

  useEffect(() => {
    // we want to connect events that are
    // related so we only generate an uuid on mount
    setTrackingUID(generateUUID());
  }, []);

  const onInputChange = (keyEvent) => {
    const value = keyEvent.target.value;
    trackInput(value, router.asPath, trackingUID);
    setQuery(value);
  };

  //@ts-ignore
  const [status, { conventions, entreprises }] = useSearchCC(query);

  return (
    <>
      <Label htmlFor="convention-search">
        Renseignez le nom de votre convention collective, le nom de votre
        entreprise ou son SIRET.
      </Label>
      <BlockInput
        placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
        value={query}
        type="search"
        name="q"
        id="convention-search"
        onChange={onInputChange}
      />
      {query && (
        <ResultsContainer>
          {status === Status.LOADING && (
            <Paragraph noMargin>
              <Spinner /> Recherche des convention collectives...
            </Paragraph>
          )}
          {status === Status.ERROR && (
            <Paragraph noMargin>
              Le service de recherche est indisponible.
            </Paragraph>
          )}
          {status === Status.EMPTY && (
            <Paragraph noMargin>Aucun résultat n’a été trouvé.</Paragraph>
          )}
          {status === Status.SUCCESS && (
            <>
              {conventions.length !== 0 && (
                <ResultList
                  title="CONVENTIONS COLLECTIVES"
                  query={query}
                  items={conventions.map((convention, index) => (
                    <ConventionLink
                      convention={convention}
                      isFirst={index === 0}
                      key={convention.slug}
                      onClick={onSelectConvention}
                    />
                  ))}
                />
              )}
              {entreprises.length !== 0 && (
                <ResultList
                  title="ENTREPRISES"
                  query={query}
                  items={entreprises.map((entreprise) => (
                    <CompanyTile
                      {...entreprise}
                      key={entreprise.siret}
                      onClick={onSelectConvention}
                    />
                  ))}
                />
              )}
            </>
          )}
        </ResultsContainer>
      )}
      <P>
        Vous ne connaissez pas ou ne trouvez pas votre convention
        collective&nbsp;?
        <Span>
          Consultez{" "}
          <HelpModal>
            {(openModal) => (
              <HelpButton
                type="button"
                variant="navLink"
                onClick={() => {
                  matopush([
                    "trackEvent",
                    "cc_search_help",
                    router.asPath,
                    `${trackingUID}`,
                  ]);
                  openModal();
                }}
              >
                notre aide
              </HelpButton>
            )}
          </HelpModal>
          .
        </Span>
      </P>
    </>
  );
};

const { spacings } = theme;
const BlockInput = styled(Input)`
  width: 100%;
`;

const P = styled.p`
  margin-top: ${spacings.xmedium};
`;

const Span = styled.span`
  display: block;
  margin-top: ${spacings.base};
`;

const HelpButton = styled(Button)`
  color: ${({ theme }) => theme.secondary};
  text-decoration: underline;
`;

const ResultsContainer = styled.div`
  margin-top: ${spacings.medium};
`;

export default Search;
