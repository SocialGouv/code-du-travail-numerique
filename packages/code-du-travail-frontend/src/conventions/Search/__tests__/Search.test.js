import { fireEvent, render, waitFor } from "@testing-library/react";
import React from "react";

import Search from "../";

global.fetch = jest.fn();
jest.useFakeTimers();

jest.mock("../../../piwik", () => ({
  matopush: jest.fn(),
}));

function renderSearchForm({
  title = "Recherche de convention collective",
  onSelectConvention = jest.fn(),
} = {}) {
  return render(
    <Search title={title} onSelectConvention={onSelectConvention} />
  );
}

const mockFetch = (resultsByService) => {
  fetch.mockReset();
  fetch.mockImplementation((url) => {
    const mockFetchResponse = (data) =>
      Promise.resolve({ json: () => Promise.resolve(data), ok: true });

    for (const [service, data] of Object.entries(resultsByService)) {
      if (new RegExp(service).test(url)) {
        return mockFetchResponse(data);
      }
    }
    return Promise.resolve({ ok: false });
  });
};

describe("<Search />", () => {
  it("should render", () => {
    const { container } = renderSearchForm({});
    expect(container).toMatchSnapshot();
  });

  it("should show spinner when loading", () => {
    mockFetch({ "api.url/idcc": {} });
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention: null,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "8888" } }
    );
    expect(container).toMatchSnapshot();
  });

  it("when input is IDCC, should search conventions", async () => {
    mockFetch({
      "api.url/idcc": {
        hits: {
          hits: [
            {
              _source: {
                id: "KALICONT000000000001",
                num: 275,
                shortTitle: "small titre",
                slug: "slug-convention-1",
                title: "titre convention 1",
              },
            },
            {
              _source: {
                id: "KALICONT000000000002",
                num: 276,
                shortTitle: "smaller convention 2",
                slug: "slug-convention-2",
                title: "titre convention 2",
              },
            },
          ],
        },
      },
    });
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention: null,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "1234" } }
    );
    jest.runAllTimers();
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));
    expect(fetch).toHaveBeenCalledWith("api.url/idcc?q=1234");
    expect(container).toMatchSnapshot();
  });

  it("when input is a valid IDCC, should show only the perfect match", async () => {
    mockFetch({
      "api.url/idcc": {
        hits: {
          hits: [
            {
              _source: {
                id: "KALICONT000000000001",
                num: 275,
                shortTitle: "small titre",
                slug: "slug-convention-1",
                title: "titre convention 1",
              },
            },
            {
              _source: {
                id: "KALICONT000000000002",
                num: 4567,
                shortTitle: "smaller convention 2",
                slug: "slug-convention-2",
                title: "titre convention 2",
              },
            },
          ],
        },
      },
    });
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention: null,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "4567" } }
    );
    jest.runAllTimers();
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));
    expect(fetch).toHaveBeenCalledWith("api.url/idcc?q=4567");
    expect(container).toMatchSnapshot();
  });

  it("should show no results when no result", async () => {
    mockFetch({
      "api.url/idcc": {
        hits: {
          hits: [],
        },
      },
    });
    const { container, getByPlaceholderText, findByText } = renderSearchForm({
      onSelectConvention: null,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "9999" } }
    );
    jest.runAllTimers();
    await waitFor(() => findByText(/Aucun résultat/i), { container });
    expect(container).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith("api.url/idcc?q=9999");
  });

  it("should use onSelectConvention callback when given", async () => {
    mockFetch({
      "api.url/idcc": {
        hits: {
          hits: [
            {
              _source: {
                id: "KALICONT000000000001",
                num: 275,
                shortTitle: "small titre",
                slug: "slug-convention-1",
                title: "titre convention 1",
              },
            },
            {
              _source: {
                id: "KALICONT000000000002",
                num: 4567,
                shortTitle: "smaller convention 2",
                slug: "slug-convention-2",
                title: "titre convention 2",
              },
            },
          ],
        },
      },
    });
    const onSelectConvention = jest.fn();
    const { getByPlaceholderText, getByText } = renderSearchForm({
      onSelectConvention,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "42" } }
    );
    jest.runAllTimers();
    await waitFor(() => expect(onSelectConvention).toHaveBeenCalledTimes(0));
    const link = getByText("smaller convention 2");
    fireEvent.click(link);
    expect(onSelectConvention).toHaveBeenCalledTimes(1);
    expect(onSelectConvention).toHaveBeenCalledWith({
      id: "KALICONT000000000002",
      num: 4567,
      shortTitle: "smaller convention 2",
      slug: "slug-convention-2",
      title: "titre convention 2",
    });
  });

  it("when searching by text, should use ES, api SIREN and siret2idcc", async () => {
    mockFetch({
      "api-entreprises.url": {
        etablissement: [
          {
            code_postal: 93100,
            id: 199308119,
            libelle_commune: "Commune test",
            nom_raison_sociale: "Entreprise 1",
            siret: "44144914700020",
          },
        ],
      },
      "api.url/idcc": {
        hits: {
          hits: [
            {
              _source: {
                id: "KALICONT000000000001",
                num: 275,
                shortTitle: "titre convention 1",
                slug: "slug-convention-1",
                title: "titre convention 1",
              },
            },
          ],
        },
      },
      "siret2idcc.url": [
        {
          conventions: [
            {
              active: true,
              date_publi: "1976-04-01T00:00:00.000Z",
              etat: "VIGUEUR_ETEN",
              id: "KALICONT000005635886",
              nature: "IDCC",
              num: 843,
              shortTitle: "Short Convention 1",
              title: "Convention 1",
            },
          ],
          siret: "44144914700020",
        },
      ],
    });

    const onSelectConvention = jest.fn();
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "hello" } }
    );
    jest.runOnlyPendingTimers(); // run debounce timer
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(3));
    expect(container).toMatchSnapshot();
  });

  it("should not use siret2idcc when no entreprise result", async () => {
    mockFetch({
      "api.url/idcc": {
        hits: {
          hits: [
            {
              _source: {
                id: "KALICONT000000000001",
                num: 275,
                shortTitle: "small title 1",
                slug: "slug-convention-1",
                title: "titre convention 1",
              },
            },
          ],
        },
      },
      "entreprise.data.gouv.fr": {
        message: "no results found",
      },
    });

    const onSelectConvention = jest.fn();
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      { target: { value: "xxxx" } }
    );
    jest.runOnlyPendingTimers(); // run debounce timer
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(2));
    expect(container).toMatchSnapshot();
  });

  it("when searching SIRET, should use API SIREN and siret2idcc", async () => {
    mockFetch({
      "api-entreprises": {
        etablissement: {
          code_postal: 93100,
          id: 199308119,
          libelle_commune: "Commune test",
          nom_raison_sociale: "Entreprise 1",
          siret: "01234567891011",
        },
      },
      "siret2idcc.url": [
        {
          conventions: [
            {
              active: true,
              date_publi: "1976-04-01T00:00:00.000Z",
              etat: "VIGUEUR_ETEN",
              id: "KALICONT000005635886",
              nature: "IDCC",
              num: "843",
              title: "Convention 1",
            },
          ],
          siret: "01234567891011",
        },
      ],
    });

    const onSelectConvention = jest.fn();
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      {
        target: { value: "01234567891011" },
      }
    );
    jest.runOnlyPendingTimers(); // run debounce timer
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(2));
    expect(container).toMatchSnapshot();
  });

  it("when searching SIREN, should use API SIREN and siret2idcc", async () => {
    mockFetch({
      "api-entreprises": {
        unite_legale: {
          denomination: "Entreprise 2",
          etablissements: [
            {
              code_postal: 93100,
              id: 199308120,
              libelle_commune: "Commune test 1",
              siret: "01234567891011",
            },
            {
              code_postal: 93120,
              etat_administratif: "F",
              id: 199308121,
              libelle_commune: "Commune test 2",
              siret: "01234567891012",
            },
          ],
        },
      },
      "siret2idcc.url": [
        {
          conventions: [
            {
              active: true,
              date_publi: "1976-04-01T00:00:00.000Z",
              etat: "VIGUEUR_ETEN",
              id: "KALICONT000005635886",
              nature: "IDCC",
              num: "843",
              title: "Convention 1",
            },
          ],
          siret: "01234567891011",
        },
        {
          conventions: [],
          siret: "01234567891012",
        },
      ],
    });

    const onSelectConvention = jest.fn();
    const { container, getByPlaceholderText } = renderSearchForm({
      onSelectConvention,
    });
    fireEvent.change(
      getByPlaceholderText(
        "Nom de la convention collective, de l’entreprise ou son SIRET"
      ),
      {
        target: { value: "012345678" },
      }
    );
    jest.runOnlyPendingTimers(); // run debounce timer
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(2));
    expect(container).toMatchSnapshot();
  });
});
