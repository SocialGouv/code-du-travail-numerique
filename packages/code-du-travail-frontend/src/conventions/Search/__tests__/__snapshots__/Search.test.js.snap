// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value=""
    />
  </span>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;

exports[`<Search /> should show no results when no result 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value="9999"
    />
  </span>
  <div
    class="sc-dycYrt lbBSWD"
  >
    <p
      class="sc-jGKxIK eoUrZA"
    >
      Aucun résultat n’a été trouvé.
    </p>
  </div>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;

exports[`<Search /> should show spinner when loading 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value="8888"
    />
  </span>
  <div
    class="sc-dycYrt lbBSWD"
  >
    <p
      class="sc-jGKxIK eoUrZA"
    >
      <svg
        aria-labelledby="title desc"
        class="__react-svg-spinner_circle"
        height="1em"
        role="img"
        style="animation-duration: 750ms;"
        viewBox="0 0 32 32"
        width="1em"
      >
        <title
          id="title"
        >
          Circle loading spinner
        </title>
        <desc
          id="desc"
        >
          Image of a partial circle indicating "loading."
        </desc>
        <style>
          
      .__react-svg-spinner_circle{
          transition-property: transform;
          animation-name: __react-svg-spinner_infinite-spin;
          animation-iteration-count: infinite;
          animation-timing-function: linear;
      }
      @keyframes __react-svg-spinner_infinite-spin {
          from {transform: rotate(0deg)}
          to {transform: rotate(360deg)}
      }
    
        </style>
        <circle
          cx="16"
          cy="16"
          fill="none"
          r="12"
          role="presentation"
          stroke="rgba(0,0,0,0.4)"
          stroke-dasharray="43.982297150257104"
          stroke-linecap="round"
          stroke-width="4"
        />
      </svg>
       Recherche des convention collectives...
    </p>
  </div>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;

exports[`<Search /> when input is IDCC, should search conventions 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value="1234"
    />
  </span>
  <div
    class="sc-dycYrt lbBSWD"
  >
    <div
      class="sc-iGgWBj sc-kMribo cRLRjd jcFZZU"
    >
      <h3
        class="sc-kAyceB dvmaTz"
        data-testid="heading"
      >
        CONVENTIONS COLLECTIVES (2
         
        résultats)
      </h3>
      <ul
        class="sc-cfxfcM kMmTaI"
      >
        <li
          class="sc-bdOgaJ gOKFpY"
        >
          <a
            class="sc-iXzfSG bmDAzy"
            href="/convention-collective/275-small-titre"
          >
            small titre
             
            <span
              class="sc-lnPyaJ VaVQQ"
            >
              (IDCC 
              0275
              )
            </span>
          </a>
        </li>
        <li
          class="sc-bdOgaJ gOKFpY"
        >
          <a
            class="sc-iXzfSG iJIMYc"
            href="/convention-collective/276-smaller-convention-2"
          >
            smaller convention 2
             
            <span
              class="sc-lnPyaJ VaVQQ"
            >
              (IDCC 
              0276
              )
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;

exports[`<Search /> when input is a valid IDCC, should show only the perfect match 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value="4567"
    />
  </span>
  <div
    class="sc-dycYrt lbBSWD"
  >
    <div
      class="sc-iGgWBj sc-kMribo cRLRjd jcFZZU"
    >
      <h3
        class="sc-kAyceB dvmaTz"
        data-testid="heading"
      >
        CONVENTIONS COLLECTIVES (1
         
        résultat)
      </h3>
      <ul
        class="sc-cfxfcM kMmTaI"
      >
        <li
          class="sc-bdOgaJ gOKFpY"
        >
          <a
            class="sc-iXzfSG bmDAzy"
            href="/convention-collective/4567-smaller-convention-2"
          >
            smaller convention 2
             
            <span
              class="sc-lnPyaJ VaVQQ"
            >
              (IDCC 
              4567
              )
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;

exports[`<Search /> when searching by text, should use enterprise API 1`] = `
<div>
  <label
    class="sc-ecPEgm jyWEST"
    for="convention-search"
  >
    Renseignez le nom de votre convention collective, le nom de votre entreprise ou son SIRET.
  </label>
  <span
    class="sc-dkmUuB ayeCS sc-empnci dzkWPn"
  >
    <input
      class="sc-ejfMa-d dijxYa"
      id="convention-search"
      name="q"
      placeholder="Nom de la convention collective, de l’entreprise ou son SIRET"
      type="search"
      value="hello"
    />
  </span>
  <div
    class="sc-dycYrt lbBSWD"
  >
    <p
      class="sc-jGKxIK eoUrZA"
    >
      Le service de recherche est indisponible.
    </p>
  </div>
  <p
    class="sc-fThUAz bqJWec"
  >
    Vous ne connaissez pas ou ne trouvez pas votre convention collective ?
    <span
      class="sc-SrznA cZVzFV"
    >
      Consultez
       
      <button
        class="sc-kOPcWz cOwJar sc-czkgLR dIDeLX"
        type="button"
      >
        notre aide
      </button>
      .
    </span>
  </p>
</div>
`;
